Fix compilation with OpenSSL < 1.1.0 and >= 0.9.8b

File src/crypto.c contains a definition of EVP_CIPHER_CTX_new() and EVP_CIPHER_CTX_free() if the module gets compiled against OpenSSL older than 1.1.0. But those two symbols have been introduced in OpenSSL 0.9.8b. For any version starting with 0.9.8b and older than 1.1.0 the compilation fails, because there are two definitions of the same symbols:

```
/path/to/openssl/lib/libcrypto.a(evp_enc.o): In function `EVP_CIPHER_CTX_new':
/path/to/openssl-src/crypto/evp/evp_enc.c:87: multiple definition of `EVP_CIPHER_CTX_new'
.libs/mod_auth_gssapi_la-crypto.o:/path/to/mod_auth_gssapi-1.6.1/src/crypto.c:31: first defined here
/path/to/openssl/lib/libcrypto.a(evp_enc.o): In function `EVP_CIPHER_CTX_free':
/path/to/openssl-src/crypto/evp/evp_enc.c:551: multiple definition of `EVP_CIPHER_CTX_free'
.libs/mod_auth_gssapi_la-crypto.o:/path/to/mod_auth_gssapi-1.6.1/src/crypto.c:44: first defined here
collect2: error: ld returned 1 exit status
```

The following patch fixes this:

```
--- src/crypto.c 2017-08-07 16:36:55.735998000 +0200
+++ src/crypto.c 2018-09-12 15:05:05.584623000 +0200
@@ -26,7 +26,9 @@
     HMAC_CTX_cleanup(ctx);
     OPENSSL_free(ctx);
 }
+#endif

+#if OPENSSL_VERSION_NUMBER < 0x0090802fL || defined(LIBRESSL_VERSION_NUMBER)
 EVP_CIPHER_CTX *EVP_CIPHER_CTX_new(void)
 {
     EVP_CIPHER_CTX *ctx;
```

--- a/configure.ac
+++ b/configure.ac
@@ -87,10 +87,16 @@ AS_IF([test "x${APR}" != "x" -a -x "${APR}"],
        AC_SUBST(APR_LDFLAGS)],
       [AC_MSG_FAILURE(["apr-1-config not found. Use --with-apr"])])
 
+# OpenSSL availability and presence of specific functions
 PKG_CHECK_MODULES([OPENSSL], [openssl])
 AC_SUBST([OPENSSL_CFLAGS])
 AC_SUBST([OPENSSL_LIBS])
+save_LIBS=$LIBS
+LIBS="$LIBS $OPENSSL_LIBS"
+AC_CHECK_FUNCS(EVP_CIPHER_CTX_new HMAC_CTX_new)
+LIBS=$save_LIBS
 
+# GSSAPI availability and presence of specific functions
 AC_CHECK_HEADERS([gssapi/gssapi.h gssapi/gssapi_ext.h gssapi/gssapi_krb5.h],
                  ,[AC_MSG_ERROR([Could not find GSSAPI headers])])
 AC_CHECK_HEADERS([gssapi/gssapi_ntlmssp.h])

--- a/src/crypto.c
+++ b/src/crypto.c
@@ -1,12 +1,13 @@
 /* Copyright (C) 2014 mod_auth_gssapi contributors - See COPYING for (C) terms */
 
+#include "config.h"
 #include <openssl/evp.h>
 #include <openssl/hmac.h>
 #include <openssl/rand.h>
 #include <stdbool.h>
 #include "crypto.h"
 
-#if OPENSSL_VERSION_NUMBER < 0x10100000L || defined(LIBRESSL_VERSION_NUMBER)
+#ifndef HAVE_HMAC_CTX_NEW
 HMAC_CTX *HMAC_CTX_new(void)
 {
     HMAC_CTX *ctx;
@@ -26,7 +27,9 @@ void HMAC_CTX_free(HMAC_CTX *ctx)
     HMAC_CTX_cleanup(ctx);
     OPENSSL_free(ctx);
 }
+#endif
 
+#ifndef HAVE_EVP_CIPHER_CTX_NEW
 EVP_CIPHER_CTX *EVP_CIPHER_CTX_new(void)
 {
     EVP_CIPHER_CTX *ctx;

